ddsc/core/filedownloader.py

```diff
@@ -0,0 +1,175 @@
+        wait_for_processes(processes, int(self.file_size), progress_queue, self.watcher, self.remote_file)
+
+    def make_big_empty_file(self):
+        """
+        Write out a empty file so the workers can seek to where they should write and write their data.
+        """
+        with open(self.path, "wb") as outfile:
+            outfile.seek(int(self.file_size) - 1)
+            outfile.write(b'\0')
+
+    def make_and_start_process(self, range_start, range_end, progress_queue):
+        """
+        Create a process that will download the specified range and notify progress_queue of progress or errors.
+        :param range_start: int: file offset to download
+        :param range_end: int: file ending offset to download
+        :param progress_queue: ProgressQueue: queue to notify as we make progress
+        :return: Process: the process we created
+        """
+        http_headers = {'Range': 'bytes={}-{}'.format(range_start, range_end)}
+        if self.http_headers:
+            http_headers.update(self.http_headers)
+        seek_amt = range_start
+        process = Process(target=download_async,
+                          args=(self.host + self.url, http_headers, self.path, seek_amt, progress_queue))
+        process.start()
+        return process
+
+
+def download_async(url, headers, path, seek_amt, progress_queue):
+    """
+    Called in separate process to download a chunk of a file.
+    :param url: str: url to file we should download
+    :param headers: dict: header to use with url, should contain Range to limit what we download
+    :param path: str: path to where we should save our chunk we download
+    :param seek_amt: int: offset to seek before writing our chunk out to path
+    :param progress_queue: ProgressQueue: queue of tuples we will add progress/errors to
+    :return:
+    """
+    downloader = ChunkDownloader(url, headers, path, seek_amt, progress_queue)
+    downloader.run()
+
+
+class ChunkDownloader(object):
+    """
+    Downloads part of a file and writes it to a location in a local pre-existing file.
+    This runs in a separate process from the main application.
+    """
+    def __init__(self, url, http_headers, path, seek_amt, progress_queue):
+        """
+        Setup for downloading part of a file.
+        :param url: str: url to the file
+        :param http_headers: dict: headers for use with the url should contain byte Range
+        :param path: str: path to file to write data to
+        :param seek_amt: int: offset amount to seek into the file
+        :param progress_queue: ProgressQueue: queue we notify of progress or errors
+        """
+        self.url = url
+        self.http_headers = http_headers
+        self.path = path
+        self.seek_amt = seek_amt
+        self.progress_queue = progress_queue
+
+    def run(self):
+        try:
+            response = requests.get(self.url, headers=self.http_headers, stream=True)
+            with open(self.path, 'r+b') as outfile:
```

Comment:
Perhaps worth a comment that 'w' means writing, but truncates the file so you have to use 'r+'. I didn't know this until I went to https://docs.python.org/2/library/functions.html#open, as I assume you did too :)